rust/arrow/src/compute/cast_kernels.rs

```diff
@@ -0,0 +1,543 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+//! Defines cast kernels for `ArrayRef`.
+//!
+//! Allows casting arrays between supported datatypes.
+//!
+//! ## Behavior
+//!
+//! * Boolean to Utf8: `true` => '1', `false` => `0`
+//! * Utf8 to numeric: strings that can't be parsed to numbers return null, float strings
+//!   in integer casts return null
+//! * Numeric to boolean: 0 returns `false`, any other value returns `true`
+//!
+//! ## Unsupported Casts
+//!
+//! * To or from `StructArray`
+//! * To or from `ListArray`
+//! * Boolean to float
+//! * Utf8 to boolean
+//!
+//! Example:
+//!
+//! ```
+//! use arrow::array::*;
+//! use arrow::compute::cast;
+//! use arrow::datatypes::DataType;
+//! use std::sync::Arc;
+//!
+//! let a = Int32Array::from(vec![5, 6, 7]);
+//! let array = Arc::new(a) as ArrayRef;
+//! let b = cast(&array, &DataType::Float64).unwrap();
+//! let c = b.as_any().downcast_ref::<Float64Array>().unwrap();
+//! assert_eq!(5.0, c.value(0));
+//! assert_eq!(6.0, c.value(1));
+//! assert_eq!(7.0, c.value(2));
+//! ```
+
+use std::sync::Arc;
+
+use crate::array::*;
+use crate::builder::*;
+use crate::datatypes::*;
+use crate::error::{ArrowError, Result};
+
+/// Macro rule to cast between numeric types
+macro_rules! cast_numeric_arrays {
```

Comment:
why these have to be macro? can we use generic function?

Reply:
I've had this problem in a lot of areas across the project. I don't know how I'd convert `DateTime::Int32` to `Int32Type`. Do you have an idea/example of how I could be able to make the cast generic?

Reply:
Can you just use:
```rust
fn cast_numeric_arrays<FROM, TO>(src: &ArrayRef) -> Result<ArrayRef>
where
    FROM: ArrowNumericType,
    TO: ArrowNumericType,
    FROM::Native: num::NumCast,
    TO::Native: num::NumCast,
{
    match numeric_cast::<FROM, TO>(
        src.as_any().downcast_ref::<PrimitiveArray<FROM>>().unwrap(),
    ) {
        Ok(to) => Ok(Arc::new(to) as ArrayRef),
        Err(e) => Err(e),
    }
}
```
This is very similar to the `numeric_cast` you have below.

Reply:
Thanks Chao, I've removed all macros and replaced with the above